pintc/src/predicate/analyse/type_check/validate.rs

```diff
@@ -236,4 +237,47 @@ impl Contract {
 
         handler.result(())
     }
+
+    pub(in crate::predicate::analyse) fn check_uninferrable_types(
+        &mut self,
+        handler: &Handler,
+    ) -> Result<(), ErrorEmitted> {
+        fn check_any_type(ty: &Type, handler: &Handler) {
+            match ty {
+                Type::Any(span) => {
+                    handler.emit_err(Error::Compile {
+                        error: CompileError::UninferrableType { span: span.clone() },
+                    });
+                }
```

Comment:
Do you have an example of this happening?  It'd be way better if we could just catch these in the type checker rather than in a validation pass.

Reply:
I understand if we're comparing `nil` to `nil` then we can't infer a type... but then the type should just be `Nil`?  Otherwise, if there is any use of the storage value at all then it should infer that type.

Reply:
Alright so yeah switched the type to `Nil` instead of `Any` in this case, but we still can't claim that `Nil == Nil` because we don't have size information. Right now, there are other cases where the type of `Nil` is not inferable such as `constraint `[ storage::a, storage::b ] == [ nil, nil ]` but that's something we could fix.

Reply:
OK.  Since the type system is actually pretty simple and we're not doing 'difficult' type inference, it would seem a bit lame for use to give up and say we couldn't infer the type.  Perhaps for very specifc `Nil` related errors we can have Errors which explain exactly why the type is unknown, and even hint how to fix it.